### PR TITLE
Edit user profile

### PR DESCRIPTION
**Pull request by markdamobi**
This pr is related to these tasks: 
https://trello.com/c/cBpCZ56s/55-profiledesign-implementation-as-a-user-i-want-to-be-able-to-edit-the-information-in-my-profile
https://trello.com/c/MBNIEj1e/57-profileheader-as-a-user-i-want-to-be-able-to-have-site-level-context-and-be-able-to-edit-my-profile-on-when-on-the-profile-page

edit_profile

- note: save functionality isn't being implemented as part of this pr. Cancel button works fine though. 
- I added this 'Edit Profile' button to the profile page so you can click through from there to get here.
- this pr contains 2 migrations. First migration to add location to user. 2nd migration to change name field on user to first_name and last_name.

### CODE REVIEW

**Comment by markdamobi** [comment]
Ignore the icon in the firstname field in my screenshot, that's just lastpass.

File: app/assets/javascripts/users/profile_edit.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+(function(){
+  $(document).ready(function(){
+    setDonationBars();
+  });
+  function setDonationBars(){
+    $(".donations-bars > .donation-bar-container").each(function(index){
+      _that = $(this);
+      var container_width = $(".donations-bars").width();
+      _that.css("width", _that.data("percent") * 1.0 * 75 / 100 + "%");
```
Comment (sanghun89) [inline app/assets/javascripts/users/profile_edit.js]: This works, but it's not a good practice to mix arithmetic logic mixed with string concatenation. We should try to abstract away to a different method, to where it looks something like:
```
_that.css("width", computeCssPercent(_that.data) + "%");
```

File: app/views/users/profile_edit/_donation_bar.html.erb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<div class="donation-bar-container" data-cause="<%= cause %>" data-percent="<%= percent %>" data-is-highest="<%= highest %>">
```
Comment (sanghun89) [inline app/views/users/profile_edit/_donation_bar.html.erb]: I don't think you have to end this element with `-container`, and rename to just `donation-bar`, as `container` classes in css is more used as a "general" purpose to refer to reusable CSS classes

File: app/views/users/profile_edit/_donations_by_causes.html.erb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+<%= render "users/profile_sections/profile_section", title: "DONATIONS BY CAUSES"  do %>
+  <div class="donations-bars">
```
Comment (sanghun89) [inline app/views/users/profile_edit/_donations_by_causes.html.erb]: we could just rename this to just `donations`. You also don't have any styling for this class so we shouldn't need it

File: app/views/users/profile_edit/_favorite_organization.html.erb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<% org_type = user_fav_org.organization.org_type %>
+<div class="favorite-organization fav-org">
```
Comment (sanghun89) [inline app/views/users/profile_edit/_favorite_organization.html.erb]: why have `favorite-organization` and `fav-org`?

File: app/views/users/profile_edit/_favorite_organizations.html.erb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<%= render "users/profile_sections/profile_section", title: "FAVORITE ORGANIZATIONS"  do %>
+  <div class="user-fav-orgs">
```
Comment (sanghun89) [inline app/views/users/profile_edit/_favorite_organizations.html.erb]: So for CSS classes, it's recommended to not have bunch of singular vs plural naming to reduce possible "one-off" bugs. Can we make some of those naming more specific? For example if they are list of things you can be like `user-favorite-orgs-list`

File: app/views/users/profile_edit/_donations_by_causes.html.erb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    <% if @donations_by_causes.present? %>
+      <div class="left-arrow" >
+        <%= fa_icon "arrow-circle-left" %>
+      </div>
+
+      <% h_cause, h_percent = @donations_by_causes[0] %>
+      <%= render "users/profile_edit/donation_bar", cause: h_cause, percent: h_percent, highest: true %>
+
```
Comment (sanghun89) [inline app/views/users/profile_edit/_donations_by_causes.html.erb]: I feel like lines 8-9 and 15-16 should be helper methods that will better explain what exactly is being done here.

something like `render_first_donation` and `render_rest_of_donations` something in that level to abstract and have better comprehension to what is being done in the view

File: app/models/user.rb [hunk 7]
```diff
@@ -7,11 +7,16 @@ class User < ApplicationRecord
 
   validates :fb_id, uniqueness: true, presence: true #this can be removed when some other signup/login method is added.
   validates :email, uniqueness: true
+  validates_presence_of :first_name, :last_name
 
   def badges
     donated_causes.uniq
   end
 
+  def name
```
Comment (arincon9) [inline app/models/user.rb]: I know it seems silly, but I think we should write a spec for `name`. I believe all public methods should be specced. There are others on this model that snuck past, but those are outside the scope of this P.R.

File: app/views/users/profile_edit/_donations_by_causes.html.erb [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+      <div class="right-arrow" >
+        <%= fa_icon "arrow-circle-right" %>
+      </div>
+
+      <% @donations_by_causes[1..-1].each do |cause, percent| %>
```
Comment (arincon9) [inline app/views/users/profile_edit/_donations_by_causes.html.erb]: Why do we need `[1..-1]`?

File: config/initializers/assets.rb [hunk 9]
```diff
@@ -12,3 +12,5 @@
 # application.js, application.css, and all non-JS/CSS in the app/assets
 # folder are already added.
 # Rails.application.config.assets.precompile += %w( admin.js admin.css )
+Rails.application.config.assets.precompile += %w( users/profile.js )
+Rails.application.config.assets.precompile += %w( users/profile_edit.js )
```
Comment (arincon9) [inline config/initializers/assets.rb]: Can we simplify this to one line: `Rails.application.config.assets.precompile += %w( users/profile.js users/profile_edit.js )`?

File: db/migrate/20190323223752_add_first_name_and_last_name_to_users.rb [hunk 10]
```diff
@@ -0,0 +1,10 @@
+class AddFirstNameAndLastNameToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column    :users, :first_name, :string
+    add_column    :users, :last_name,  :string
+    remove_column :users, :name,       :string
+
+    add_index     :users, :first_name
+    add_index     :users, :last_name
```
Comment (arincon9) [inline db/migrate/20190323223752_add_first_name_and_last_name_to_users.rb]: For future reference, we could have also done `add_index :users, [:first_name, :last_name]`

Comment (markdamobi) [reply db/migrate/20190323223752_add_first_name_and_last_name_to_users.rb]: That's not quite what I wanted. They don't give the same thing.

File: config/initializers/assets.rb (hunk 9 above)
Comment (markdamobi) [reply config/initializers/assets.rb]: yeah, I'll take care of that.

File: app/models/user.rb (hunk 7 above)
Comment (markdamobi) [reply app/models/user.rb]: You're right. It skipped my mind. Thanks for pointing it out. I'll take care of it.

File: app/views/users/profile_edit/_favorite_organizations.html.erb (hunk 5 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: Can you explain more? I don't see what's wrong with `user-fav-orgs`. I'm also not sure what you mean by "one-off" bugs. Can you give an example?

File: app/views/users/profile_edit/_favorite_organization.html.erb (hunk 4 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organization.html.erb]: Mostly just for flexibility.

File: app/views/users/profile_edit/_donations_by_causes.html.erb (hunk 3 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_donations_by_causes.html.erb]: I have some styling for it. Look in `users/profile_edit/donations_by_causes.scss`. I don't `donations` would be a good name for it because it isn't a group of donations. Each item is not a donation. I think that might be a bit misleading.

File: app/views/users/profile_edit/_donation_bar.html.erb (hunk 2 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_donation_bar.html.erb]: I use container because it contains the bar, the cause text and the percentage. Maybe I can rename the partial to `donation_bar_container`. What do you think?
There's actually a possibility of this container being re-used. I was considering re-using it on the profile page instead of what I currently have there. But that's not part of the scope of this pr.

File: app/assets/javascripts/users/profile_edit.js (hunk 1 above)
Comment (markdamobi) [reply app/assets/javascripts/users/profile_edit.js]: Yeah, good point. I can take care of that.

File: app/assets/javascripts/users/profile_edit.js [hunk 11]
```diff
@@ -0,0 +1,12 @@
+(function(){
+  $(document).ready(function(){
+    setDonationBars();
+  });
+  function setDonationBars(){
+    $(".donations-bars > .donation-bar-container").each(function(index){
+      _that = $(this);
```
[the remark this answers is not in the document]
Comment (markdamobi) [reply app/assets/javascripts/users/profile_edit.js]: Thanks for these. I'll work on it. I don't really know most js conventions so these are helpful.

File: app/views/users/profile_edit/_donations_by_causes.html.erb (hunk 6 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_donations_by_causes.html.erb]: I can look into this.

File: app/views/users/profile_edit/_favorite_organizations.html.erb (hunk 5 above)
Comment (sanghun89) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: Not to be confused with traditional "off-by-one" errors, I mean to point out the fact that CSS is the number one place to look at wrong selectors. If the selector names are very similar, it's easy to debug or develop on wrong classes.  It's easier to catch on general application languages, but CSS tend to be really painful when one thing is named `user-fav-org` (singular) and another being `user-fav-orgs` (plural).

File: app/views/users/profile_edit/_favorite_organization.html.erb (hunk 4 above)
Comment (sanghun89) [reply app/views/users/profile_edit/_favorite_organization.html.erb]: Flexibility??? what do you mean?

File: app/views/users/profile_edit/_donations_by_causes.html.erb (hunk 3 above)
Comment (sanghun89) [reply app/views/users/profile_edit/_donations_by_causes.html.erb]: So for CSS, you have to be cognitive on what we are **seeing**, as CSS is applied to what we actually see. From the view, it literally says "Donation By Causes", which can fairly point out that this div element is being applied with a `donations` class tag.

However, looking further into it, something like `donation-distribution` might make more sense.

Comment (markdamobi) [reply app/views/users/profile_edit/_donations_by_causes.html.erb]: I'll probably go with `donation-distribution`. I think that sounds good.

File: app/views/users/profile_edit/_favorite_organizations.html.erb (hunk 5 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: I don't think confusing singular and plural is much of a concern here. This can be said about a lot of other things. But if you feel very strongly about this, I can change it to `user-fav-orgs-list`.

File: app/views/users/profile_edit/_favorite_organization.html.erb (hunk 4 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organization.html.erb]: Some people might like the whole thing being spelled out. Others may prefer the short form.

File: app/views/users/profile_edit/_donation_bar.html.erb (hunk 2 above)
Comment (sanghun89) [reply app/views/users/profile_edit/_donation_bar.html.erb]: So when I mean reusability I meant the literal class name.

`container` is a term heavily used in bootstrap in where if you apply the class `container` in other elements, it applies basic margin, padding and maybe few other things to get the grid going. 

You will probably not going to apply `donation-bar-container` class to other elements. If you're talking about this whole `erb` layout, sure we can use that multiple times, but not the actual class names. We're probably not going to be in a word where we have
```
<div class="profile donation-bar-container">
```

File: app/views/users/profile_edit/_favorite_organizations.html.erb (hunk 5 above)
Comment (sanghun89) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: I'm only coming from experience, having done many many frontend applications in the past.

CSS is not really a programming language. You can think of it as using tags and depend on them to apply some changes. If tags are way too similarly named, there's a chance of misusing them

File: app/views/users/profile_edit/_favorite_organization.html.erb (hunk 4 above)
Comment (sanghun89) [reply app/views/users/profile_edit/_favorite_organization.html.erb]: let's just go with one I personally think it creates more confusion to implement something that's not needed

File: app/assets/stylesheets/users/profile_edit/action_buttons.scss [hunk 12]
```diff
@@ -0,0 +1,23 @@
+.user-profile-edit__form {
+  .action-buttons {
```
Comment (sanghun89) [inline app/assets/stylesheets/users/profile_edit/action_buttons.scss]: typically when you are doing BEM, you declare these at top level, meaning it shouldn't be underneath `.user-profile-edit__form`

File: app/views/users/profile_edit/_donation_bar.html.erb (hunk 2 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_donation_bar.html.erb]: alright, cool. I can change it.

File: app/views/users/profile_edit/_favorite_organization.html.erb (hunk 4 above)
Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organization.html.erb]: no problem.

File: app/assets/stylesheets/users/profile_edit/action_buttons.scss (hunk 12 above)
Comment (markdamobi) [reply app/assets/stylesheets/users/profile_edit/action_buttons.scss]: I want to be sure that this styling is just for the action buttons within the form. I don't want it some other place for now. How can I achieve that if I move it to top level?

File: app/helpers/users/profile_edit_helper.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  def render_remaining_donation_bars(donations_by_causes:)
+    donations_by_causes[1..-1].each do |cause, percent|
+      concat(render_donation_bar(cause: cause, percent: percent))
+    end
+  end
+
+  def render_donation_bar(cause:, percent:, highest: false)
+    render "users/profile_edit/donation_bar", cause: cause, percent: percent, highest: highest
+  end
+end
```
Comment (sanghun89) [inline app/helpers/users/profile_edit_helper.rb]: can you add tests for these as well??

File: app/views/users/profile_edit/_favorite_organizations.html.erb [hunk 14]
```diff
@@ -0,0 +1,7 @@
+<%= render "users/profile_sections/profile_section", title: "FAVORITE ORGANIZATIONS"  do %>
+  <div class="user-fav-orgs-list">
+    <% @user.user_favorite_organizations[0..3].each do |user_fav_org| %> <!-- show only 4 favorite organizations for now. -->
```
Comment (sanghun89) [inline app/views/users/profile_edit/_favorite_organizations.html.erb]: Sorry I didn't see this either. I think we should similarly add this to one of helpers and also make pulling the top `number` as a constant variable.

Also test for this as well..

Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: I don't think this one needs to be moved to any helper right now. What it's doing is quite minimal and it doesn't seem confusing. I can always refactor it eventually if/when more logic is added and things get more complicated.

Comment (sanghun89) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: This applies to the same reason as my request to other helpers. These are to abstract away logic in layout to provide better readability and testability, which is something we should strive for and demonstrate as software engineers

Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: I’ll be happy to look more into this on a different task that more focused on this section. I think it’s okay for now.

Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: I’ll be happy to look more into this on a different task that more focused on this section. I think it’s okay for now.

Comment (sanghun89) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: I'm not sure why this portion is "different". This is an implementation that relates to two trello stories you worked from, and the change request is almost identical to requests I had to something that's very similar to this. 

I would be willing to fallback if I'm really pushing a difficult refactor, but this is something that will improve the view implementation that is in this PR.

Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: For this one, I really don't think a helper is needed for this one. But I can remove the extra iteration logic though. I'll pass in a collection to the partial instead. I'll make the change and you let me what you think.

Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: > This is an implementation that relates to two trello stories you worked

Not sure what you mean.

Comment (markdamobi) [reply app/views/users/profile_edit/_favorite_organizations.html.erb]: @sanghun89 I've removed some of the logic from the view. Take another look at it.
There's probably other places that logic can be extracted from. But I still don't think it's within the scope of the pr to extract all the view logic from all places. I'll be happy to focus on that in subsequent tasks.

File: db/migrate/20190323223752_add_first_name_and_last_name_to_users.rb (hunk 10 above)
Comment (arincon9) [reply db/migrate/20190323223752_add_first_name_and_last_name_to_users.rb]: Oops my bad
